### PR TITLE
some file scanner performance improvements

### PR DESCRIPTION
**Pull request by icewind1991**
Some pretty low hanging optimizations that improve scanner performance by ~20% in my testing

### CODE REVIEW

**Review by artonge — APPROVED** [review]
Looks good.

Just to be sure, this does the following:

- cache `$blacklist` in `FileSystem::isFileBlacklisted` + use it in more places
- cache `$config` and `mimeTypeDetector` in `Local`. Why can't we use DI here?
- cache file's `full_path` in `Local->stat`

File: lib/private/Files/Filesystem.php [hunk 1]
```diff
@@ -492,9 +494,12 @@ public static function isBlacklisted($data) {
 	public static function isFileBlacklisted($filename) {
 		$filename = self::normalizePath($filename);
 
-		$blacklist = \OC::$server->getConfig()->getSystemValue('blacklisted_files', ['.htaccess']);
+		if (!self::$blacklist) {
```
Comment (come-nc) [inline lib/private/Files/Filesystem.php:497]:
```suggestion
		if (self::$blacklist === null) {
```
Blacklist may be an empty array, no?

**Comment by icewind1991** [comment]
> use it in more places

More moving around where it's used to prevent double checking the same path

> Why can't we use DI here?

DI is currently not possible in `Storage` backends

**Review by come-nc — APPROVED** [review]

File: lib/private/Files/Filesystem.php [hunk 2]
```diff
@@ -66,6 +66,8 @@ class Filesystem {
 
 	private static $listeningForProviders = false;
 
+	private static $blacklist = null;
```
Comment (CarlSchwan) [inline lib/private/Files/Filesystem.php:69]:
```suggestion
	private static array $blacklist = null;
```

**Review by CarlSchwan — APPROVED** [review]

File: lib/private/Files/Filesystem.php (hunk 2 above)
Comment (icewind1991) [reply lib/private/Files/Filesystem.php:69]: doesn't work because it's `array|null`

Comment (CarlSchwan) [reply lib/private/Files/Filesystem.php:70]: private static ?array should then work

**Pull request merged**